### ISSUE TITLE
basic scheduler: limit number of cranks per block

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#2299]
## What is the Problem Being Solved?

Currently, the cosmic-swingset host application runs the kernel's run-queue to completion on each block, by calling `controller.run()` during the `endBlock` phase:

https://github.com/Agoric/agoric-sdk/blob/07e763d4749fcbc1252178a93f09a9e253632942/packages/cosmic-swingset/lib/launch-chain.js#L96-L99

If the run-queue is too deep (either because an inbound message added too much, the processing of one queue entry causes too many additional entries to be added, or because the queue was too deep to start with), the block proposer will take too long to generate the proposal, and the other validators will give up on it. Whatever time the block proposer takes to create the block will also be required in the validators to check it.

Currently, a simple infinite loop spread across two separate vats will cause an unending stream of events to be added to the run-queue, one at a time, and `controller.run()` will never finish.

Instead of an unbounded `controller.run()`, the host application needs to call `controller.step()` a finite number of times (and/or we should enhance swingset to offer `controller.run(maxCranks)` and cosmic-swingset should call it with a finite argument).

For now, we should pick any finite value (perhaps `1000`) for `maxCranks`: anything is better than allowing an infinite loop. If the number is too low, we may leave work on the queue and under-utilize the CPU. If the number is too high, we may exceed our block-processing (time) budget and validators may get slashed.

In the longer term, we need to find a way to make this adaptive. Ideally, we would compute a wallclock-time processing budget as `(INTER_BLOCK_TIME - MESSAGING_AND_VOTING_OVERHEAD) / 2 - SAFETY_MARGIN`, and have the block proposer execute as many cranks as it could within that budget. That would utilize as much of the CPU as possible without causing votes or blocks to be delayed (and achieving roughly 50% CPU utilization on each node).

The tricky thing is that we don't know ahead of time how long each crank will take. Our Metering story may provide some rough upper bounds, but it is still likely to be variable.

I'll file a separate ticket to brainstorm about how to make this adaptive. For now, the main task is to allow a finite number of cranks.

Note that anything other than "inifinity" creates the scarceness that makes Scheduling interesting. If all cranks are executed on every block, there's no value to prioritizing some messages over others. But if only a subset can be executed, then paying for priority gets you a clear benefit. So meaningful Scheduler/Escalator work is dependent upon this limitation being in place.

# Test Plan

We should have some unit tests in cosmic-swingset that can exercise smaller parts of the system.

In the longer run, we want a performance testnet with a client that creates periodic bursts of traffic (more than 1000 cranks), but whose overall rate is low enough to allow that traffic to drain. We should be able to observe the later messages in the burst taking more than one block time to get handled.

cc @michaelfig

**Comment by michaelfig** [comment]
I like the spirit of this.  However, one detail doesn't jive with my understanding of the Tendermint/Cosmos-SDK split.  The block proposer can only select a set of transactions to propose.  It actually does the work of running the block manager (BEGIN_BLOCK, DELIVER_INBOUND, END_BLOCK, COMMIT_BLOCK, etc) at the same time as the other validators do, which is after the proposal has passed voting and achieved consensus.

The proposer doesn't have a valid Cosmos SDK context in which to evaluate transactions and process the run queue until after the block's contents have been agreed to.  The app hash of those evaluated transactions only becomes part of the next proposed block.

This model is simpler than the one explained above, as there is not any timing dependency between the proposer finishing evaluating transactions and the other validators beginning evaluating the same transactions.  I think the Tendermint folk would not want to complicate their app vs. consensus layering, and I don't think we'll be in a position to start Agoric-specific innovation in this space for the foreseeable future.

**Comment by warner** [comment]
Yeah, I don't think I understand the cosmos/tendermint model fully. If the proposer executes the full set of transactions before proposing a block (so the block can contain an apphash that matches/includes/derives-from the same block's transactions), then the available processing time is limited to half the block time, since the proposer and validators cannot run in parallel. If the block's apphash is derived from the txns in the *previous* block, then the proposer of block-1 is committing the proposer of block-2 to some amount of work (which seems like maybe an attack vector), however then a proposer and all the validators might get to run in parallel.

If I understand what you're saying, the sequence looks like:

* proposer of block-1 chooses a set of transactions, publishes the txset-1 in a proposed block-1
* validators receive proposed block-1, deliberate, vote
* proposed block receives enough votes
* all nodes (proposer and validators) process txset-1 from block-1, come up with apphash-1
* ... ? do all nodes then exchange apphash-1 and vote on it? I don't think so, which means it must be included in the next block
* proposer of block-2 chooses the next set of txns, publishes txset-2 and apphash-1 as proposed block-2

Which I guess matches the second approach in my first paragraph.

So if it had followed the first approach, my bright idea was for the block proposer to generate synthetic "run one crank" txns, one at a time. We'd have cosmic-swingset run `c.step()` for each one. The proposer would pay attention to the wallclock time after each step, and stop producing them when the budget was exceeded. Then it publishes a block that includes some number of "real" (IO) txns, and some number of RunOneCrank txns, and all followers run the same number of cranks (so they ought to finish within their time budget too).

If it's using the second approach, we could still have the block producer emit a "run N cranks" txn to tell everyone else how much work to do, but it couldn't know how long that would take ahead of time, so it wouldn't know a good value of N to use. And it's spending all its budget executing the *previous* block's txns.

Hm, and it wouldn't help to give each node an extra CPU core, because we can't speculatively execute txset-2 (to measure how long it takes, to figure out how many txns to include in txset-2) without first finishing execution of txset-1.

If all nodes took precisely the same amount of wallclock time (impossible), or if our execution model has precise metering (more likely) and that metering was linearly related to wallclock time (less likely), then we could circulate a gas limit, and rely upon all nodes individually choosing the exact same number of cranks to run within that limit (or just slightly past that limit, really). We'd need a mechanism to adjust that gas limit over time, to effectively learn the seconds-per-gas ratio and maximize utilization. Hm, this is not a new problem, Ethereum has a form of it, and I'm sure Cosmos has put a lot of thought into it.

Hrmph. I don't have a good feel for the solution yet.

**Comment by michaelfig** [comment]
[I was confused by thinking "block-1" was "block minus one".]

> If the block's apphash is derived from the txns in the _previous_ block, then the proposer of block-1 is committing the proposer of block-2 to some amount of work (which seems like maybe an attack vector), however then a proposer and all the validators might get to run in parallel.

The proposer of block1 is committing **all** of the validators of block2 to the same amount of work.  If 2/3 of them complete the prevote, then if some cannot complete the work before the commitment timeout, those ones have missed a block.  It's okay now and then, but missing too many (a tunable parameter) means being jailed and slashed for downtime.

Attempting to link the gas limit (which is one of the parameters in the proposal) to the execution time seems to be the way other chains solve it, then the validators can vote against a proposal that has too high of a gas budget.  I agree that the metering way of determining how many cranks to run (and having that be a governance-controlled swingset parameter) is the simplest way I can think of to extend that protection beyond Tendermint to the VM.  (I had a flicker of that idea when we were talking in my post-demoday haze.)

**Comment by michaelfig** [comment]
My suggestion for a scheduling algorithm would be:

1. Start with a block meter budget of `(n + 1) * m`, where `m` is our maximum meter per crank and `n` is the minimum number of cranks per block.  `m` and `n` must be chosen so that the block meter budget limits the worst-case wallclock time to our tolerance.
2. Run a crank with a vat meter of `m`.  The vat is terminated if it consumes its entire meter.
3. Deduct the actual meter amount that the crank took from the block budget.
4. If the kernel run queue is nonempty and the remaining block meter budget is more than `m`, go back to step 2
5. Commit the block

**Comment by warner** [comment]
I'm moving the general scheduling design discussion to #2319, leaving this ticket for the narrowly-scoped immediate task to apply any sort of consistent limit to each block.
